File: src/main/java/brightspark/runicmagic/init/RMSpells.java
package brightspark.runicmagic.init;

import brightspark.runicmagic.enums.RuneType;
import brightspark.runicmagic.spell.Spell;
import brightspark.runicmagic.spell.projectile.SpellBolt;
import brightspark.runicmagic.spell.projectile.SpellConfuse;
import brightspark.runicmagic.spell.projectile.SpellStrike;
import brightspark.runicmagic.spell.projectile.SpellWeaken;
import brightspark.runicmagic.spell.self.*;
import brightspark.runicmagic.spell.teleport.SpellTeleportGatestone;
import brightspark.runicmagic.spell.teleport.SpellTeleportHome;
import brightspark.runicmagic.spell.teleport.SpellTeleportSpawn;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.registries.IForgeRegistry;

import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

// https://runescape.fandom.com/wiki/List_of_spells
public class RMSpells
{
	public static IForgeRegistry<Spell> REGISTRY;

	public static void register(IForgeRegistry<Spell> registry)
	{
		registry.registerAll(
				//Teleport Spells
				new SpellTeleportSpawn(),
				new SpellTeleportHome(),
				new SpellGatestoneCreate(),
				new SpellTeleportGatestone(),

				//Projectile Attack Spells
				new SpellStrike(RuneType.AIR, 1),
				new SpellStrike(RuneType.WATER, 5),
				new SpellStrike(RuneType.EARTH, 9),
				new SpellStrike(RuneType.FIRE, 13),
				new SpellBolt(RuneType.AIR, 17),
				new SpellBolt(RuneType.WATER, 23),
				new SpellBolt(RuneType.EARTH, 29),
				new SpellBolt(RuneType.FIRE, 35),

				//Projectile Effect Spells
				new SpellWeaken(),
				new SpellConfuse(),

				//Self Spells
				new SpellEnchant(1, 7).addRuneCost(RuneType.WATER, 1),
				new SpellEnchant(2, 27).addRuneCost(RuneType.AIR, 3),
				new SpellEnchant(3, 49).addRuneCost(RuneType.FIRE, 5),
				new SpellEnchant(4, 57).addRuneCost(RuneType.EARTH, 10),
				new SpellEnchant(5, 68).addRuneCost(RuneType.EARTH, 15).addRuneCost(RuneType.WATER, 15),
				new SpellEnchant(6, 87).addRuneCost(RuneType.FIRE, 20).addRuneCost(RuneType.EARTH, 20),
				new SpellBonesToApples(),
				new SpellHumidify(),

				//Charge Orb
				new SpellChargeOrb(RuneType.WATER, 56),
				new SpellChargeOrb(RuneType.EARTH, 60),
				new SpellChargeOrb(RuneType.FIRE, 63),
				new SpellChargeOrb(RuneType.AIR, 66)
		);
	}

	public static List<String> getRegNames()
	{
		return REGISTRY.getKeys().stream()
			.map(ResourceLocation::toString)
			.sorted(String::compareToIgnoreCase)
			.collect(Collectors.toCollection(LinkedList::new));
	}

	public static Spell getSpell(String name)
	{
		return REGISTRY.getValue(new ResourceLocation(name));
	}

	public static List<Spell> getSortedSpells()
	{
		List<Spell> spells = new LinkedList<>(REGISTRY.getValuesCollection());
		spells.sort(Comparator.comparingInt(Spell::getLevel));
		return spells;
	}
}

File: src/main/kotlin/brightspark/runicmagic/init/RMSpells.kt
package brightspark.runicmagic.init

import brightspark.runicmagic.RunicMagic
import brightspark.runicmagic.item.RuneItem
import brightspark.runicmagic.model.RuneType.*
import brightspark.runicmagic.model.SpellType
import brightspark.runicmagic.spell.Spell
import brightspark.runicmagic.spell.projectile.ProjectileBaseSpell
import brightspark.runicmagic.spell.self.*
import brightspark.runicmagic.spell.teleport.GatestoneTeleportSpell
import brightspark.runicmagic.spell.teleport.HomeTeleportSpell
import brightspark.runicmagic.util.setRegName
import net.minecraft.item.Item
import net.minecraft.util.ResourceLocation
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.registries.IForgeRegistry
import net.minecraftforge.registries.RegistryBuilder
import java.awt.Color

// https://runescape.fandom.com/wiki/List_of_spells
object RMSpells {
	// TODO: Move this somewhere better?
	private val EARTH_COLOUR: Color = Color(0x00BB00)

	lateinit var REGISTRY: IForgeRegistry<Spell>

	val REGISTRY_KEYS: Iterable<ResourceLocation> by lazy { REGISTRY.keys.sorted() }

	fun createRegistry() {
		REGISTRY = RegistryBuilder<Spell>()
			.setName(ResourceLocation(RunicMagic.MOD_ID, "spells"))
			.setType(Spell::class.java)
			.disableSaving()
			.allowModification()
			.create()
	}

	fun register(event: RegistryEvent.Register<Spell>) = event.registry.registerAll(
		// Teleport
		spell("teleport_home", HomeTeleportSpell(teleportProps(SpellType.TELESELF, 40))),
		spell("teleport_gatestone", GatestoneTeleportSpell(teleportProps(SpellType.TELESELF, 32))),

		// Projectile Attack
		// FYI: Damage values are from Runescape - we scale it down in #elementalSpell
		elementalSpell("strike_air", 1, AIR.colour, 48),
		elementalSpell("strike_water", 5, WATER.colour, 48),
		elementalSpell("strike_earth", 9, EARTH_COLOUR, 86),
		elementalSpell("strike_fire", 13, FIRE.colour, 124),
		elementalSpell("bolt_air", 17, AIR.colour, 163),
		elementalSpell("bolt_water", 23, WATER.colour, 220),
		elementalSpell("bolt_earth", 29, EARTH_COLOUR, 278),
		elementalSpell("bolt_fire", 35, FIRE.colour, 336),
		elementalSpell("blast_air", 41, AIR.colour, 393),
		elementalSpell("blast_water", 47, WATER.colour, 451),
		elementalSpell("blast_earth", 53, EARTH_COLOUR, 508),
		elementalSpell("blast_fire", 59, FIRE.colour, 566),
		elementalSpell("wave_air", 62, AIR.colour, 595),
		elementalSpell("wave_water", 65, WATER.colour, 624),
		elementalSpell("wave_earth", 70, EARTH_COLOUR, 672),
		elementalSpell("wave_fire", 75, FIRE.colour, 720),
		elementalSpell("surge_air", 81, AIR.colour, 777),
		elementalSpell("surge_water", 85, WATER.colour, 816),
		elementalSpell("surge_earth", 90, EARTH_COLOUR, 864),
		elementalSpell("surge_fire", 95, FIRE.colour, 883),

		// Projectile Effect

		// Self
		enchantSpell("enchant_1", 5, enchantProps(7).addRuneCost(COSMIC to 1, WATER to 1)),
		enchantSpell("enchant_2", 10, enchantProps(27).addRuneCost(COSMIC to 1, AIR to 3)),
		enchantSpell("enchant_3", 15, enchantProps(49).addRuneCost(COSMIC to 1, FIRE to 5)),
		enchantSpell("enchant_4", 20, enchantProps(57).addRuneCost(COSMIC to 1, EARTH to 10)),
		enchantSpell("enchant_5", 25, enchantProps(68).addRuneCost(COSMIC to 1, EARTH to 15, WATER to 15)),
		enchantSpell("enchant_6", 30, enchantProps(87).addRuneCost(COSMIC to 1, EARTH to 20, FIRE to 20)),
		spell(
			"bones_to_apples",
			BonesToApplesSpell(
				alchemyProps(15).setCastTime(40).setCooldown(1200).addRuneCost(NATURE to 1, EARTH to 2, WATER to 2)
			)
		),
		spell(
			"humidify",
			HumidifySpell(alchemyProps(68).setCastTime(80).addRuneCost(ASTRAL to 1, FIRE to 1, WATER to 3))
		),
		spell(
			"gatestone_create",
			GatestoneCreateSpell(
				selfProps(SpellType.OTHER, 32).setCastTime(60).setCooldown(600).addRuneCost(COSMIC to 3)
			)
		),

		// Charge Orb
		chargeOrbSpell("charge_orb_water", RMItems.ORB_WATER, enchantProps(56).addRuneCost(COSMIC to 3, WATER to 30)),
		chargeOrbSpell("charge_orb_earth", RMItems.ORB_EARTH, enchantProps(60).addRuneCost(COSMIC to 3, EARTH to 30)),
		chargeOrbSpell("charge_orb_fire", RMItems.ORB_FIRE, enchantProps(63).addRuneCost(COSMIC to 3, FIRE to 30)),
		chargeOrbSpell("charge_orb_air", RMItems.ORB_AIR, enchantProps(66).addRuneCost(COSMIC to 3, AIR to 30))
	)

	private fun props(spellType: SpellType, level: Int): Spell.Properties = Spell.Properties(spellType, level)

	private fun teleportProps(spellType: SpellType, level: Int): Spell.Properties =
		props(spellType, level).setSelectable(false).setCooldown(600).setCastTime(200)

	// TODO: Add cast time to projectile spells for an animation
	private fun projectileProps(spellType: SpellType, level: Int): Spell.Properties =
		props(spellType, level).setSelectable(true).setCooldown(10)

	private fun elementalProps(level: Int): Spell.Properties = projectileProps(SpellType.ELEMENTAL, level)

	private fun selfProps(spellType: SpellType, level: Int): Spell.Properties =
		props(spellType, level).setSelectable(false)

	private fun alchemyProps(level: Int): Spell.Properties = selfProps(SpellType.ALCHEMY, level)

	private fun enchantProps(level: Int): Spell.Properties = selfProps(SpellType.ENCHANTMENT, level)

	private fun spell(name: String, spell: Spell): Spell = spell.setRegName(name)

	// FYI: Damage atm is calculated as Runescape spell damage / 25
	private fun elementalSpell(name: String, level: Int, colour: Color, damage: Int): Spell =
		spell(name, ProjectileBaseSpell(elementalProps(level), RMEntities.SPELL, colour, damage.toFloat() / 25F))

	private fun enchantSpell(name: String, enchantLevel: Int, props: Spell.Properties): Spell =
		spell(name, EnchantSpell(enchantLevel, props))

	private fun chargeOrbSpell(name: String, runeItem: Item, props: Spell.Properties): Spell =
		spell(name, ChargeOrbSpell(runeItem as RuneItem, props))

	fun get(regName: String): Spell? = REGISTRY.getValue(ResourceLocation(regName))

	fun get(regName: ResourceLocation): Spell? = REGISTRY.getValue(regName)

	fun getAllSorted(): List<Spell> = REGISTRY.values.sortedBy { it.level }
}
